### PR TITLE
Rejigger Cell Visibility Tracking

### PR DESCRIPTION
**Pull request by Adlai-Holler**
cc #145  . I never had a reliable repro for it, so I'm counting on the community to test this patch.

`willDisplayCell` and `didEndDisplayingCell` are the authoritative sources for what elements are visible. Elements may be visible in multiple cells e.g. during animations, so we use a counted set.

- Replace `(weak)_ASCollectionViewCell.node` with `(strong)_ASCollectionViewCell.element`.
- Add an `NSCountedSet<ASCollectionElement *> *_visibleElements` to collection view.
- In `willDisplayCell:` and `didEndDisplayingCell:`, add/remove the cell's element from the visible set.
- Modify `visibleElementsForRangeController:` to just return a copy of this set.
- Put back the dealloc-while-visible assertion.

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by Adlai-Holler** [comment]
@rurza @temkich @appleguy Could y'all test this patch and see if it resolves the dealloc-while-visible cases you were seeing? It should be solid.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/7155529d359792b8f84e3095325335c2/raw/7b7be474d1962e851077dd2cfbb035f71cd25acc/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/ASCollectionView.mm [hunk 1]
```diff
@@ -1031,30 +1037,33 @@ - (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cell
...
   }
 
   // Since _ASCollectionViewCell is not available for subclassing, this is faster than isKindOfClass:
   // We must exit early here, because only _ASCollectionViewCell implements the -node accessor method.
-  if ([cell class] != [_ASCollectionViewCell class]) {
+  if ([rawCell class] != [_ASCollectionViewCell class]) {
     [_rangeController setNeedsUpdate];
     return;
   }
+  auto cell = (_ASCollectionViewCell *)rawCell;
```
Comment (appleguy) [inline Source/ASCollectionView.mm]: y hallo there, auto

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by rurza** [comment]
Super :) Ok
Spark by Readdle

**Comment by Adlai-Holler** [comment]
Let's land this and get it tested. I'll keep an eye on its behavior and fix any regressions.

**Pull request merged**